lib/Backend/Vault/Kv.hs

```diff
@@ -0,0 +1,231 @@
+import           Servant.Client.Core.Response (responseStatusCode)
+import           Servant.Client.Core.Request  (RequestF (Request))
+import           Network.HTTP.Types           (statusCode)
+
+import           Polysemy
+import           Control.Lens
+
+data VaultKvBackend =
+  VaultKvBackend
+  { _vbName :: T.Text
+  , _vbAddress :: BaseUrl
+  , _vbMount :: T.Text
+  , _vbToken :: I.VaultToken
+  }
+  deriving (Show)
+
+{-# INLINE didimatch #-}
+didimatch
+    :: (b -> Maybe a)  -- ^ Mapper for consumer
+    -> (a -> Maybe b)     -- ^ Mapper for producer
+    -> TomlCodec a  -- ^ Source 'Codec' object
+    -> TomlCodec b  -- ^ Target 'Codec' object
+didimatch matchB matchA codec = Toml.Codec
+    { Toml.codecRead = \t -> case Toml.codecRead codec t of
+        Success a -> maybe empty Success (matchA a)
+        Failure b -> Failure b
+    , Toml.codecWrite = \c -> case matchB c of
+        Nothing -> empty
+        Just d  -> Toml.codecWrite codec d >>= maybe empty pure . matchA
+    }
+
+
+vaultKvCodec :: TomlCodec VaultKvBackend
+vaultKvCodec = VaultKvBackend
+                  <$> Toml.text "name" Toml..= _vbName
+                  <*> didimatch baseUrlToText textToBaseUrl (Toml.text "address") Toml..= _vbAddress
+                  <*> Toml.text "mount" Toml..= _vbMount
+                  <*> didimatch tokenToText textToToken (Toml.text "token") Toml..= _vbToken
+  where tokenToText (I.VaultToken t) = Just t
+        textToToken t = Just $ I.VaultToken t
+        baseUrlToText = Just . T.pack . showBaseUrl
+        textToBaseUrl = parseBaseUrl . T.unpack
+
+data FieldMetadata = FieldMetadata
+  { _dateModified :: UTCTime
+  , _private :: Bool
+  }
+  deriving stock (Show, Generic)
+  deriving anyclass (A.ToJSON, A.FromJSON)
+makeLenses ''FieldMetadata
+
+data CofferSpecials =
+  CofferSpecials
+  { _masterKey :: Maybe T.Text
+  , _globalDateModified :: UTCTime
+  , _fields :: HS.HashMap T.Text FieldMetadata
+  , _tags :: [T.Text]
+  }
+  deriving stock (Show, Generic)
+  deriving anyclass (A.ToJSON, A.FromJSON)
+makeLenses ''CofferSpecials
+
+runVaultIO :: Member (Embed IO) r
+           => Member (Error CofferError) r
+           => BaseUrl
+           -> I.VaultToken
+           -> T.Text
+           -> Sem (BackendEffect ': r) a
+           -> Sem r a
+runVaultIO url token mount = interpret $
+  \x -> do
+    env <-
+      case url of
+        (BaseUrl Http _ _ _) -> do
+          manager <- embed $ newManager defaultManagerSettings
+          pure $ mkClientEnv manager url
+        (BaseUrl Https _ _ _) -> do
+          manager <- embed $ newManager tlsManagerSettings
+          pure $ mkClientEnv manager url
+
+    case x of
+      WriteSecret entry -> do
+        let
+          cofferSpecials = CofferSpecials
+            { _masterKey =
+              entry ^. E.masterField <&> E.getFieldKey
+            , _globalDateModified = entry ^. E.dateModified
+            , _fields =
+              entry ^. E.fields & traverse %~
```

Comment:
When using `lens`, `each` is more idiomatic than `traverse`. We should do that here (and in a few other places where we're using `traverse` in a lens expression).

Reply:
Done